File: packages/javascript/drtrace-client/src/cli/grep.ts
/**
 * Grep command implementation for searching logs with POSIX regex.
 * Matches Python implementation in packages/python/src/drtrace_service/cli/grep.py
 */

import * as fs from 'fs';
import * as path from 'path';

interface GrepOptions {
  pattern: string;
  ignoreCase?: boolean;
  count?: boolean;
  invertMatch?: boolean;
  lineNumber?: boolean;
  extendedRegex?: boolean;
  since?: string;
  applicationId?: string;
  daemonHost?: string;
  daemonPort?: number;
  json?: boolean;
  color?: 'auto' | 'always' | 'never';
}

interface LogRecord {
  id: number;
  ts: number;
  level: string;
  message: string;
  application_id: string;
  service_name?: string;
  module_name?: string;
  file_path?: string;
  line_no?: number;
  exception_type?: string | null;
  stacktrace?: string | null;
  context: Record<string, any>;
}

/**
 * Check if daemon is available with timeout
 */
async function checkDaemonAlive(host: string, port: number, timeoutMs: number = 500): Promise<boolean> {
  const controller = new AbortController();
  const timeout = setTimeout(() => controller.abort(), timeoutMs);
  
  try {
    const response = await fetch(`http://${host}:${port}/status`, {
      signal: controller.signal,
    });
    clearTimeout(timeout);
    return response.ok;
  } catch (error) {
    clearTimeout(timeout);
    return false;
  }
}

/**
 * Format timestamp for display
 */
function formatTimestamp(ts: number): string {
  const date = new Date(ts * 1000);
  return date.toISOString().replace('T', ' ').substring(0, 19);
}

/**
 * Apply color to text based on log level
 */
function colorize(text: string, level: string, colorMode: 'auto' | 'always' | 'never'): string {
  const shouldColor = colorMode === 'always' || (colorMode === 'auto' && process.stdout.isTTY);
  
  if (!shouldColor) {
    return text;
  }
  
  const colors: Record<string, string> = {
    ERROR: '\x1b[31m',    // Red
    CRITICAL: '\x1b[31m', // Red
    WARNING: '\x1b[33m',  // Yellow
    WARN: '\x1b[33m',     // Yellow
    INFO: '\x1b[0m',      // Default
    DEBUG: '\x1b[90m',    // Gray
  };
  
  const reset = '\x1b[0m';
  const color = colors[level] || colors.INFO;
  
  return `${color}${text}${reset}`;
}

/**
 * Main grep implementation
 */
export async function grep(options: GrepOptions): Promise<number> {
  const {
    pattern,
    ignoreCase = false,
    count = false,
    invertMatch = false,
    lineNumber = false,
    extendedRegex = false,
    since = '5m',
    applicationId,
    daemonHost = process.env.DRTRACE_DAEMON_HOST || 'localhost',
    daemonPort = parseInt(process.env.DRTRACE_DAEMON_PORT || '8001'),
    json = false,
    color = 'auto',
  } = options;
  
  // Check if daemon is available
  const daemonAvailable = await checkDaemonAlive(daemonHost, daemonPort, 500);
  
  if (!daemonAvailable) {
    console.error('Error: DrTrace daemon is not running.');
    console.error('');
    console.error('Start the daemon with:');
    console.error('  drtrace daemon start');
    return 2;
  }
  
  // Build query parameters (Story 11-2: wire -E flag to message_regex)
  const params = new URLSearchParams({
    since: since,
  });
  
  // Use message_regex if -E flag provided, else message_contains (Epic 11.1, 11.2)
  if (extendedRegex) {
    params.append('message_regex', pattern);
  } else {
    params.append('message_contains', pattern);
  }
  
  if (applicationId) {
    params.append('application_id', applicationId);
  }
  
  // Query daemon using fetch
  try {
    const daemonUrl = `http://${daemonHost}:${daemonPort}/logs/query`;
    const response = await fetch(`${daemonUrl}?${params.toString()}`);
    
    if (!response.ok) {
      const error = await response.json() as { detail?: { message?: string } | string };
      const errorMsg = typeof error.detail === 'object' ? error.detail?.message : error.detail;
      console.error(`Error: ${errorMsg || 'Query failed'}`);
      return 2;
    }
    
    const data = await response.json() as { results?: LogRecord[] };
    const results: LogRecord[] = data.results || [];
    
    // Apply additional filters that weren't sent to API (invert_match)
    let filteredResults = results;
    
    if (invertMatch) {
      if (extendedRegex) {
        const flags = ignoreCase ? 'i' : '';
        const regex = new RegExp(pattern, flags);
        filteredResults = results.filter(record => !regex.test(record.message));
      } else {
        const searchPattern = ignoreCase ? pattern.toLowerCase() : pattern;
        filteredResults = results.filter(record => {
          const message = ignoreCase ? record.message.toLowerCase() : record.message;
          return !message.includes(searchPattern);
        });
      }
    }
    
    // Output results
    if (count) {
      console.log(filteredResults.length);
      return 0;
    }
    
    if (json) {
      console.log(JSON.stringify(filteredResults, null, 2));
      return 0;
    }
    
    if (filteredResults.length === 0) {
      return 1;
    }
    
    // Format and print results
    for (const record of filteredResults) {
      const tsStr = formatTimestamp(record.ts);
      const serviceStr = record.service_name ? `[${record.service_name}]` : '';
      let msg = `[${tsStr}] ${serviceStr} [${record.level}] ${record.message}`;
      
      msg = colorize(msg, record.level, color);
      
      if (lineNumber) {
        // Use timestamp as pseudo line number for daemon results
        msg = `${Math.floor(record.ts)}:${msg}`;
      }
      
      console.log(msg);
    }
    
    return 0;
  } catch (error: unknown) {
    const errorMessage = error instanceof Error ? error.message : String(error);
    console.error(`Error: Failed to query daemon: ${errorMessage}`);
    return 2;
  }
}

/**
 * Parse CLI arguments and run grep command
 */
export async function runGrep(args: string[]): Promise<number> {
  // Simple argument parser matching Python argparse behavior
  const options: GrepOptions = {
    pattern: '',
    ignoreCase: false,
    count: false,
    invertMatch: false,
    lineNumber: false,
    extendedRegex: false,
    since: '5m',
    color: 'auto',
  };
  
  let i = 0;
  while (i < args.length) {
    const arg = args[i];
    
    if (arg === '-i' || arg === '--ignore-case') {
      options.ignoreCase = true;
    } else if (arg === '-c' || arg === '--count') {
      options.count = true;
    } else if (arg === '-v' || arg === '--invert-match') {
      options.invertMatch = true;
    } else if (arg === '-n' || arg === '--line-number') {
      options.lineNumber = true;
    } else if (arg === '-E' || arg === '--extended-regex') {
      options.extendedRegex = true;
    } else if (arg === '--since') {
      options.since = args[++i];
    } else if (arg === '--application-id') {
      options.applicationId = args[++i];
    } else if (arg === '--daemon-host') {
      options.daemonHost = args[++i];
    } else if (arg === '--daemon-port') {
      options.daemonPort = parseInt(args[++i]);
    } else if (arg === '--json') {
      options.json = true;
    } else if (arg === '--color') {
      const colorValue = args[++i];
      if (colorValue !== 'auto' && colorValue !== 'always' && colorValue !== 'never') {
        console.error(`Error: Invalid color value '${colorValue}'. Use: auto, always, never`);
        return 2;
      }
      options.color = colorValue;
    } else if (arg === '-h' || arg === '--help') {
      printHelp();
      return 0;
    } else if (!arg.startsWith('-')) {
      // Positional argument - pattern
      options.pattern = arg;
    } else {
      console.error(`Error: Unknown option '${arg}'`);
      return 2;
    }
    
    i++;
  }
  
  if (!options.pattern) {
    console.error('Error: Pattern is required');
    printHelp();
    return 2;
  }
  
  return grep(options);
}

function printHelp(): void {
  console.log(`Usage: drtrace grep [OPTIONS] PATTERN

Search log messages with pattern matching.

Options:
  -i, --ignore-case       Ignore case in pattern matching
  -c, --count             Output count of matches instead of matches
  -v, --invert-match      Invert match (show non-matching lines)
  -n, --line-number       Output line numbers with matches
  -E, --extended-regex    Use POSIX extended regex (sends message_regex to API)
  --since <time>          Time range: 30m/1h/2d/7d (default: 5m)
  --application-id <id>   Filter by application ID
  --daemon-host <host>    Daemon host (default: localhost)
  --daemon-port <port>    Daemon port (default: 8001)
  --json                  Output in JSON format
  --color <mode>          Color control: auto, always, never (default: auto)
  -h, --help              Show this help message

Examples:
  # Simple substring search
  drtrace grep "timeout"

  # Regex search with -E flag
  drtrace grep -E "error|warning"

  # Search with application filter
  drtrace grep --application-id myapp "error"

  # Count matches
  drtrace grep -c "error"
`);
}

File: packages/python/src/drtrace_service/cli/grep.py
"""Grep command implementation for searching logs with POSIX regex."""

import argparse
import os
import re
import sys
import time
from datetime import datetime, timedelta
from pathlib import Path
from typing import List, Optional, Tuple

import httpx

from drtrace_service.daemon_health import check_daemon_alive
from drtrace_service.storage import get_default_log_path


# Cache for recent log reads to avoid repeated disk scans
_log_cache: dict = {}
_cache_ttl: float = 30.0  # 30 seconds


def _get_cached_log(log_path: Path) -> Optional[List[str]]:
    """Get log from cache if fresh."""
    now = time.time()
    if log_path in _log_cache:
        content, timestamp = _log_cache[log_path]
        if now - timestamp < _cache_ttl:
            return content
        else:
            del _log_cache[log_path]
    return None


def _cache_log(log_path: Path, lines: List[str]) -> None:
    """Cache log file lines."""
    _log_cache[log_path] = (lines, time.time())


def _parse_time_duration(duration_str: str) -> Optional[timedelta]:
    """Parse duration string like '30m', '1h', '2d', '7d'.
    
    Returns:
        timedelta object or None if invalid
    """
    import re
    match = re.match(r'^(\d+)([mhd])$', duration_str.strip().lower())
    if not match:
        return None
    
    value, unit = int(match.group(1)), match.group(2)
    if unit == 'm':
        return timedelta(minutes=value)
    elif unit == 'h':
        return timedelta(hours=value)
    elif unit == 'd':
        return timedelta(days=value)
    return None


def _parse_log_line(line: str) -> Optional[Tuple[datetime, str, str, str]]:
    """Parse log line to extract timestamp, service, level, message.
    
    Expected format: [YYYY-MM-DD HH:MM:SS] [SERVICE] [LEVEL] MESSAGE
    
    Returns:
        Tuple of (datetime, service, level, message) or None if parse fails
    """
    import re
    pattern = r'^\[(\d{4}-\d{2}-\d{2} \d{2}:\d{2}:\d{2})\] \[([^\]]+)\] \[([^\]]+)\] (.*)$'
    match = re.match(pattern, line)
    if not match:
        return None
    
    try:
        timestamp = datetime.strptime(match.group(1), '%Y-%m-%d %H:%M:%S')
        return (timestamp, match.group(2), match.group(3), match.group(4))
    except ValueError:
        return None


def _should_include_line(
    line: str,
    pattern: str,
    ignore_case: bool,
    invert_match: bool,
    extended_regex: bool,
    since: Optional[timedelta] = None
) -> bool:
    """Determine if a line should be included based on filters.
    
    Args:
        line: Log line to test
        pattern: Regex pattern to match
        ignore_case: If True, ignore case in pattern matching
        invert_match: If True, invert match (include non-matching)
        extended_regex: If True, use POSIX extended regex
        since: If provided, only include lines after this duration
        
    Returns:
        True if line should be included, False otherwise
    """
    # Parse line if time filtering needed
    parsed = None
    if since:
        parsed = _parse_log_line(line)
        if parsed:
            timestamp, _, _, _ = parsed
            cutoff = datetime.now() - since
            if timestamp < cutoff:
                return False
    
    # Pattern matching
    flags = re.IGNORECASE if ignore_case else 0
    try:
        if extended_regex:
            # Python's re module supports extended regex
            match = re.search(pattern, line, flags)
        else:
            # Basic regex (just search, not extended)
            match = re.search(pattern, line, flags)
    except re.error:
        # Invalid regex pattern
        return False
    
    # Apply invert logic
    if invert_match:
        return not match
    else:
        return bool(match)


def grep_command(args: Optional[List[str]] = None) -> int:
    """Execute grep command.
    
    Args:
        args: Command-line arguments (for testing)
        
    Returns:
        Exit code: 0 (matches found), 1 (no matches), 2 (error)
    """
    parser = argparse.ArgumentParser(
        prog='drtrace grep',
        description='Search logs with POSIX regex'
    )
    parser.add_argument('pattern', help='Regex pattern to search for')
    parser.add_argument('-i', '--ignore-case', action='store_true', help='Ignore case in pattern')
    parser.add_argument('-c', '--count', action='store_true', help='Output count of matches instead of matches')
    parser.add_argument('-v', '--invert-match', action='store_true', help='Invert match (output non-matching lines)')
    parser.add_argument('-n', '--line-number', action='store_true', help='Output line numbers with matches')
    parser.add_argument('-E', '--extended-regex', action='store_true', help='Use POSIX extended regex')
    parser.add_argument('--since', default=None, help='Time range: 30m/1h/2d/7d')
    parser.add_argument('--full-search', action='store_true', help='Allow searches >30d without limit')
    parser.add_argument('--json', action='store_true', help='Output in JSON format')
    
    try:
        parsed_args = parser.parse_args(args)
    except SystemExit:
        return 2
    
    # Validate time range
    if parsed_args.since:
        since_td = _parse_time_duration(parsed_args.since)
        if not since_td:
            print(f"Error: Invalid time duration '{parsed_args.since}'", file=sys.stderr)
            return 2
        
        if since_td.days > 30 and not parsed_args.full_search:
            print(
                f"Error: Time range {parsed_args.since} exceeds 30 days. "
                "Use --full-search to allow longer searches.",
                file=sys.stderr
            )
            return 2
    else:
        since_td = None
    
    # Check if daemon is available
    daemon_available = check_daemon_alive(timeout_ms=500)
    
    if daemon_available:
        # Use daemon HTTP query path (Story 11-2: wire -E flag to message_regex)
        try:
            daemon_host = os.getenv("DRTRACE_DAEMON_HOST", "localhost")
            daemon_port = os.getenv("DRTRACE_DAEMON_PORT", "8001")
            daemon_url = f"http://{daemon_host}:{daemon_port}/logs/query"
            
            # Build query parameters
            params = {
                "since": parsed_args.since if parsed_args.since else "5m",
            }
            
            # Use message_regex if -E flag provided, else message_contains (Epic 11.1, 11.2)
            if parsed_args.extended_regex:
                params["message_regex"] = parsed_args.pattern
            else:
                params["message_contains"] = parsed_args.pattern
            
            # Query daemon using httpx
            try:
                response = httpx.get(daemon_url, params=params, timeout=5.0)
                response.raise_for_status()
                data = response.json()
            except (httpx.RequestError, httpx.HTTPStatusError) as e:
                # Daemon query failed, fall back to local file
                raise
            else:
                # Process daemon results
                from drtrace_service.models import LogRecord
                
                results = []
                for record_dict in data.get("results", []):
                    record = LogRecord(**record_dict)
                    
                    # Apply additional filters that weren't sent to API (invert_match, line_number)
                    if parsed_args.invert_match:
                        if parsed_args.extended_regex:
                            flags = re.IGNORECASE if parsed_args.ignore_case else 0
                            if re.search(parsed_args.pattern, record.message, flags):
                                continue
                        else:
                            pattern_lower = parsed_args.pattern.lower() if parsed_args.ignore_case else parsed_args.pattern
                            message_lower = record.message.lower() if parsed_args.ignore_case else record.message
                            if pattern_lower in message_lower:
                                continue
                    
                    # Format output
                    ts_str = datetime.fromtimestamp(record.ts).strftime('%Y-%m-%d %H:%M:%S')
                    service_str = f"[{record.service_name}]" if record.service_name else ""
                    msg = f"[{ts_str}] {service_str} [{record.level}] {record.message}"
                    
                    if parsed_args.line_number:
                        # Use timestamp as pseudo line number for daemon results
                        results.append(f"{int(record.ts)}:{msg}")
                    else:
                        results.append(msg)
                
                # Output results
                if parsed_args.count:
                    print(len(results))
                elif results:
                    for result in results:
                        print(result)
                else:
                    return 1
                
                return 0
        except Exception as e:
            # Daemon query failed, fall back to local file
            pass
    
    # Use local file fallback
    log_path = get_default_log_path()
    if not log_path or not log_path.exists():
        print(f"Error: Log file not found at {log_path}", file=sys.stderr)
        return 2
    
    # Check cache first
    lines = _get_cached_log(log_path)
    if lines is None:
        try:
            with open(log_path, 'r', encoding='utf-8') as f:
                lines = f.readlines()
            _cache_log(log_path, lines)
        except (IOError, OSError) as e:
            print(f"Error: Could not read log file: {e}", file=sys.stderr)
            return 2
    
    # Apply filters
    matches = []
    for line_num, line in enumerate(lines, start=1):
        line = line.rstrip('\n')
        if _should_include_line(
            line,
            parsed_args.pattern,
            parsed_args.ignore_case,
            parsed_args.invert_match,
            parsed_args.extended_regex,
            since_td
        ):
            if parsed_args.line_number:
                matches.append(f"{line_num}:{line}")
            else:
                matches.append(line)
    
    # Output results
    if parsed_args.count:
        print(len(matches))
    elif matches:
        for match in matches:
            print(match)
    else:
        # No matches - empty output, exit code 1
        return 1
    
    return 0
